File: qml-intro/ex-model-from-cpp/pyqt/mymodel.py
#!/usr/bin/env python3
#
# Copyright (c) 2019, Klaralvdalens Datakonsult AB (KDAB)
# All rights reserved.
#
# See the LICENSE.txt file shipped along with this file for the license.
#
#

# This Python file uses the following encoding: utf-8

import random

from PyQt5.QtCore import QAbstractListModel, QObject, QMetaEnum, QTimer, pyqtSlot as Slot, QModelIndex, QVariant


class Data:
    def __init__(self, name, flag, population):
        super(Data, self).__init__()

        self.name = name
        self.flag = flag
        self.population = population


class MyModel(QAbstractListModel, QObject):

    def __init__(self):
        super(MyModel, self).__init__()

        self._m_data = [Data("Denmark", "images/denmark.jpg", 5.6),
                        Data("Sweden", "images/sweden.jpg", 9.6),
                        Data("Iceland", "images/iceland.jpg", 0.3),
                        Data("Norway", "images/norway.jpg", 5.1),
                        Data("Finland", "images/finland.jpg", 5.4)]

        growthTimer = QTimer(self)
        growthTimer.timeout.connect(self.growPopulation)
        growthTimer.start(2000)
        self.RAND_MAX = 2147483647

    class Roles(QMetaEnum):
        NameRole = 0
        FlagRole = 1
        PopulationRole = 2

    def rowCount(self, parent):
        if parent.isValid():
            return 0
        return len(self._m_data)

    def data(self, index, role):
        if not index:
            return QVariant

        if role == self.Roles.NameRole:
            return self._m_data[index.row()].name
        elif role == self.Roles.FlagRole:
            return self._m_data[index.row()].flag
        elif role == self.Roles.PopulationRole:
            return self._m_data[index.row()].population
        else:
            return QVariant

    def roleNames(self):
        _name = "name".encode('utf-8')
        _flag = "flag".encode('utf-8')
        _population = "population".encode('utf-8')

        mapping = {self.Roles.NameRole: _name,
                   self.Roles.FlagRole: _flag,
                   self.Roles.PopulationRole: _population}
        return mapping

    @Slot(int)
    def duplicateData(self, row):
        if row < 0 | row > len(self._m_data):
            return

        data = self._m_data[row]
        rowOfInsert = row + 1

        QAbstractListModel.beginInsertRows(self, QModelIndex(), rowOfInsert, rowOfInsert)
        self._m_data.insert(rowOfInsert, data)
        QAbstractListModel.endInsertRows(self)

    @Slot(int)
    def removeData(self, row):
        if row < 0 | row > len(self._m_data):
            return

        data = self._m_data[row]

        QAbstractListModel.beginRemoveRows(self, QModelIndex(), row, row)
        self._m_data.remove(data)
        QAbstractListModel.endRemoveRows(self)

    @Slot()
    def growPopulation(self):
        growthFactor = 0.01 / self.RAND_MAX

        count = len(self._m_data)
        for i in range(0, count):
            self._m_data[i].population += self._m_data[i].population * random.randint(1, self.RAND_MAX) * growthFactor

        startIndex = QAbstractListModel.index(self, 0, 0)
        endIndex = QAbstractListModel.index(self, count - 1, 0)

        self.dataChanged.emit(startIndex, endIndex, [self.Roles.PopulationRole])

File: qml-intro/ex-model-from-cpp/cpp/mymodel.cpp
/*************************************************************************
 *
 * Copyright (c) 2010-2019, Klaralvdalens Datakonsult AB (KDAB)
 * All rights reserved.
 *
 * See the LICENSE.txt file shipped along with this file for the license.
 *
 *************************************************************************/

#include "mymodel.h"

#include <QByteArray>
#include <QTimer>
#include <cstdlib>

MyModel::MyModel(QObject *parent) :
    QAbstractListModel(parent)
{
    m_data
        << Data("Denmark", "qrc:images/denmark.jpg", 5.6)
        << Data("Sweden", "qrc:images/sweden.jpg", 9.6)
        << Data("Iceland", "qrc:images/iceland.jpg", 0.3)
        << Data("Norway", "qrc:images/norway.jpg", 5.1)
        << Data("Finland", "qrc:images/finland.jpg", 5.4);

    QTimer *growthTimer = new QTimer(this);
    connect(growthTimer, &QTimer::timeout, this, &MyModel::growPopulation);
    growthTimer->start(2000);
}

int MyModel::rowCount( const QModelIndex& parent) const
{
    if (parent.isValid())
        return 0;

    return m_data.count();
}

QVariant MyModel::data(const QModelIndex &index, int role) const
{
    if ( !index.isValid() )
        return QVariant();

    const Data &data = m_data.at(index.row());
    if ( role == NameRole ){
        return data.name;
    }
    else if ( role == FlagRole )
        return data.flag;
    else if ( role == PopulationRole )
        return data.population;
    else
        return QVariant();
}

//--> slide
QHash<int, QByteArray> MyModel::roleNames() const
{
    static QHash<int, QByteArray> mapping {
        {NameRole, "name"},
        {FlagRole, "flag"},
        {PopulationRole, "population"}
    };
    return mapping;
}
//<-- slide

void MyModel::duplicateData(int row)
{
    if (row < 0 || row >= m_data.count())
        return;

    const Data data = m_data[row];
    const int rowOfInsert = row + 1;

    beginInsertRows(QModelIndex(), rowOfInsert, rowOfInsert);
    m_data.insert(rowOfInsert, data);
    endInsertRows();
}

void MyModel::removeData(int row)
{
    if (row < 0 || row >= m_data.count())
        return;

    beginRemoveRows(QModelIndex(), row, row);
    m_data.removeAt(row);
    endRemoveRows();
}

void MyModel::growPopulation()
{
    const double growthFactor = 0.01 / RAND_MAX;

    const int count = m_data.count();
    for (int i = 0; i < count; ++i) {
        m_data[i].population += m_data[i].population * qrand() * growthFactor;
    }

    // we've just updated all rows...
    const QModelIndex startIndex = index(0, 0);
    const QModelIndex endIndex   = index(count - 1, 0);

    // ...but only the population field
    emit dataChanged(startIndex, endIndex, QVector<int>() << PopulationRole);
}
